[Inductor-CPP] If all of the activation scale dims are 1, make it a 0D tensor

sanchitintel opened the pull request:
For int8 dynamically quantized activation & int8 quantized weights, add a workaround for some indexing issue that expected an empty index ( so, was expecting a 0D tensor) in epilogue creator when the activation scale was sized [1, 1] by converting it into a 0D tensor.

The issue was discovered while running LLaMA2 quantized with torchao's `int8_dynamic_activation_int8_weight` quantization on CPU with max-autotune enabled (although this error would've occurred regardless).

The final hidden states tensor that's activation to LM head is of shape `[batch_size, sequence_length, hidden_dim]` during decoding. For decoding one token at a time with batch size 1, sequence length is 1. The activation scale is shaped `[1, 1]` (reshaped from `[1, 1, 1]`). However, Inductor epilogue creator expects a 0D tensor in this case (my guess is that the corresponding logic in Inductor expects a 0D tensor if a tensor has only one element, even if it's 1D?).






cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

sanchitintel committed:
[WIP][Inductor-CPP] If activation scale is sized [1, 1], make it a 0D tensor

For int8 dynamically quantized activation & int8 quantized weights, add a workaround for some indexing issue that expected an empty index ( so, was expecting a 0D tensor) in epilogue creator when the activation scale was sized [1, 1] by converting it into a 0D tensor.

TODO:
- [ ] Run CI. While 2D activation scales are converted into 1D, that may not always be the case

sanchitintel commented on torch/_inductor/mkldnn_lowerings.py:
```diff
@@ -889,6 +896,25 @@ def qlinear_binary(
+                    x_scale = view(x_scale, [])
+            assert len(x_scale.get_size()) in [0, 1], "x_scale must be 0D or 1D"
+
+            if x_zp is None:
+                x_zp = V.graph.add_tensor_constant(
+                    torch.tensor(0, dtype=torch.int32), name="x_zp"
+                )
+
+            if w_zp is None:
+                w_zp = V.graph.add_tensor_constant(
+                    torch.tensor(0, dtype=torch.int32), name="w_zp"
+                )
```
This particular change is now necessary because of [an unnecessary change](https://github.com/pytorch/pytorch/pull/143187#discussion_r1921052467) introduced in https://github.com/pytorch/pytorch/pull/143187/commits/09bdb53dc849735714d767fd81d59b3c402da38f

sanchitintel committed:
[skip ci] Update

sanchitintel commented:
@pytorchbot rebase -b main

sanchitintel committed:
Add UT

leslie-fang-intel commented on torch/_inductor/mkldnn_lowerings.py:
```diff
@@ -588,6 +588,14 @@ def qlinear_unary(
             else:
+                x_scale_size = x_scale.get_size()
+                if len(x_scale_size) == 2:
+                    if x_scale_size[0] == 1 and x_scale_size[1] == 1:
```
Can we generalize this check as all the value inside `x_scale_size` is 1? and how about the case when `x_scale_size` is `[1, 1, 1]`?

leslie-fang-intel commented on torch/_inductor/mkldnn_lowerings.py:
```diff
@@ -894,7 +902,26 @@ def qlinear_binary(
                 )
             else:
+                x_scale_size = x_scale.get_size()
+                if len(x_scale_size) == 2:
```
Do we have the UT to cover this case for `qlinear_binary`?

sanchitintel replied:
@leslie-fang-intel @chunyuan-w, to clarify, the change here is for fixing `qlinear_binary` lowering, which is unrelated to GEMM auto-tuning. So, it's not appropriate for me to add a UT for it in `test_cpu_select_algorithm.py`, since the GEMM template is currently not working for `qlinear_binary` with sum post-op right now. I'll add a UT in `test_mkldnn_pattern_matcher.py` instead. Thanks!

jansel commented on torch/_inductor/mkldnn_lowerings.py:
```diff
@@ -588,6 +588,14 @@ def qlinear_unary(
                     torch.tensor(x_scale, dtype=torch.float32), name="x_scale"
                 )
             else:
+                x_scale_size = x_scale.get_size()
+                if len(x_scale_size) == 2:
+                    if x_scale_size[0] == 1 and x_scale_size[1] == 1:
```
```suggestion
            if all(i==1 for in x_scale.get_size()):
```

sanchitintel replied:
Thank you! I applied this change at a couple of places.

sanchitintel committed:
[skip ci] [Change 1/2] Add UT

sanchitintel committed:
Apply jansel's review suggestion

sanchitintel committed:
Had forgotten to remove dead code :(

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -3644,6 +3646,8 @@ def forward(self, a):
+                    c.add_(self.additive)
                 return c
 
         mod = Mod(dtype, has_bias).eval()
```
If `inplace_add` is True, do we need to check that `qlinear_binary` fusion has been hit?

sanchitintel replied:
For these UTs, the Inductor pattern-matching passes in quantization.py are not responsible for lowering to `qlinear_binary`, so I couldn't use a corresponding counter, and added a check for the generated code instead.

BTW, here are some findings unrelated to the specific fix in this PR:
 
 1. `qlinear_binary` is not being used with the UTs when `M` dimension is not 1. I don't know if that's intentional. cc @Xia-Weiwen 
 2. An Inductor bug got exposed for dynamic shapes - when dynamic shapes are enabled and `x2` is added with a sum post-op in `qlinear_binary`, the outermost dim's stride of the output tensor can't be determined (The comments in code have a typo & state `innermost` instead of `outermost`). Details below.
 
 <details>

```python
# s1 below is undetermined
def call(args):
    arg4_1, arg5_1 = args
    args.clear()
    s0 = arg4_1
    assert_size_stride(arg5_1, (1, s0), (s0, 1))
    # Topologically Sorted Source Nodes: [add], Original ATen: [aten.add]
    torch.ops.onednn.qlinear_pointwise.binary_tensor(arg5_1, reinterpret_tensor(_frozen_param1, (), (), 0), x_zp, _frozen_param4, _frozen_param5, w_zp, _frozen_param3, None, 1.0, 0, torch.float32, 1.0, 0, 'sum', 1.0, 'none', [None], '')
    del arg5_1
    return (reinterpret_tensor(_frozen_param3, (1, 64), (s1, 1), 0), )
```

</details>

Xia-Weiwen replied:
> qlinear_binary is not being used with the UTs when M dimension is not 1.

You may add cases for that I think. I don't think we have restrictions on M for qllinear_binary.

sanchitintel replied:
Sorry, I meant when M > 1, then lowering to `qlinear_binary` is not done

sanchitintel committed:
Check for qlinear_binary in generated code

sanchitintel committed:
Fix lint and comments

sanchitintel committed:
Fix typo in UT

sanchitintel commented:
Hi @leslie-fang-intel, thanks for adding the `trunk` label!
Looks like I'd have to disable the UT for ARM CPU based machines

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -3657,7 +3696,10 @@ def matcher_check_fn():
             (a,),
-            matcher_check_fn=matcher_check_fn,
+            matcher_check_fn,
+            include_ops=("qlinear_pointwise.binary_tensor",)
```
Why don't we just check the `counters["inductor"]["qlinear_binary_matcher_count"]`?

sanchitintel replied:
Now that I think of it. I should've used the counter.

For `[M=32, K=32, N=64]`, this pattern wasn't hit (and not even when I changed `K` to 48, it wasn't, so the counter was 0), and I thought this lowering wasn't because of those FX patterns. When I later discovered that only the `M=1` case was being lowered to `qlinear_binary`, I should've re-added the counter check.

leslie-fang-intel commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -3632,18 +3661,28 @@ def __init__(self, dtype: torch.dtype, has_bias: bool):
                 if self.has_bias:
                     c = c + self.bias
+                elif inplace_add and test_for_pointwise_binary:
+                    # When M is 1, dynamic shapes are enabled with torch.compile, has_bias is False,
```
When M is 1, I think Inductor will always do the specialization instead of using the symbolic size.

sanchitintel replied:
Tracking that issue at #148199. Thanks!

sanchitintel committed:
Revise as per review comments

sanchitintel committed:
Fix typo & lint

sanchitintel committed:
Fix typo

leslie-fang-intel commented on test/inductor/test_cpu_select_algorithm.py:
```diff
@@ -1444,7 +1445,6 @@ def test_da8w8_sym_act_sym_wgt_with_int_mm(
-        M = 32
         in_feature = 48
         out_feature = 64
         q_min, q_max = -32, 31
```
It seems this unit test doesn't verify whether the CPP Template has been used. Can we add a counter check similar to https://github.com/pytorch/pytorch/pull/147895?

sanchitintel replied:
I could rebase this PR when #147895 will be merged (but it seems like its merge procedure ran into some CI issue), so I'll add it now.

sanchitintel committed:
Add select_algorithm_autotune counter

sanchitintel commented:
@pytorchbot rebase -b main

sanchitintel committed:
Update test_cpu_select_algorithm.py

leslie-fang-intel commented on test/inductor/test_cpu_select_algorithm.py:
```diff
@@ -1498,6 +1498,9 @@ def forward(self, a):
         self._check_amx_counter(vec_amx)
+        if torch._C._cpu._is_amx_tile_supported():
+            # Only AMX ISA based micro-kernel is currently supported for da8w8
+            self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
```
I think we should use the new counters add by you recently, right?

sanchitintel replied:
No, because #147895 didn't get merged (timed out due to non-completion of some CI jobs)

sanchitintel committed:
Merge branch 'main' into sanchitintel/fix_llama_da8w8_corner_case

sanchitintel committed:
Realize x_scale before get_size() call

Some commit on or after Feb 25 now again requires calling x_scale.realize() before a get_size() call on it
